### DISCUSSION TITLE
Cache Data on Runtime

### DISCUSSION THREAD

**Discussion by Yimura** [YimMenu/YimMenu#446] [Research]
Currently YimMenu has 2 points of failure as it's relying on data hosted by others and proxied by me over non-secure HTTP.

## The problems

1. We rely on DurtyFree to keep his gtav-data repository up-to-date. He might stop updating this repository or take the repository offline completely.
2. We rely on myself to have a proxy running that goes from HTTPS to non-secure HTTP. My server will not be available forever or the domain name / IP might change in the future.
3. The amount of files served are quickly using a bunch of bandwidth. (last month we used 4.8TB of data)
![image](https://user-images.githubusercontent.com/24669514/196745659-68a5f61d-534c-4323-867f-6ea6770256ed.png)

## Possible Solutions

1. Rehosting DurtyFree's repository will no longer makes us reliant on DurtyFree to keep the repository online, though if he stops updating it we'll have to updated it ourselves.
2. Adding a library to YimMenu that supports HTTPS will make it so we no longer need to rely on my to proxy the files. Though adding OpenSSL to the build dependencies might make it significantly harder for people to start with YimMenu as a base.
3. The bandwidth isn't a problem as of now but might be a problem in the future if it really gets out of hand, if we remove the need to proxy any files this will resolve itself.

## My suggestion

I would like to make it so YimMenu can provide this data by itself on runtime, now obviously loading this data every time again would be wasteful that's why I would like the make the following proposal.

### Caching

To cache our data I don't want to store it inside of the YimMenu DLL. Storing data like models, spawn hashes, strings, you name it has several disadvantages like being easily detectable by the signature scanner in GTA as well as just bloating the DLL.

Thus I would like to store this data to disk in a binary format, storing in binary has 2 advantages:
1. Storage Efficiency
2. Fast (no time wasted converting ASCII to integers, floats, etc...)

With the main downside being it not being very readable for humans.

This is my (up for changes) proposal for a file header / format.
|Order|Field Name|Field Size (bytes)|Type|
|--|--|--|--|
|1|cache version|4|`uint32`|
|2|game build|4|`uint32`|
|3|online version|4|`float`|
|4|data size|8|`uint64`|
|5|data| / | `any` |

### Collecting Data (@ runtime)

We can collect models using the ModelHashTable in the game and combine this with GET_LABEL_TEXT and other natives to collect the necessary data. Though this might be a heavy procedure which is why I would cache this once as long as cache version, game build, online version, ... match.

## Why am I creating this post?

Mostly just to look at how we'd best collect the data provided by DurtyFree and no longer have to rely on his repository. Since we can't collect everything from the ModelHashTable.

**Comment by maybegreat48** [comment]
Collecting vehicle info using the HashTable shouldn't be an expensive operation at all. It would even allow the menu to spawn vehicles added by SP mods. Obtaining the ped info might be very difficult however as they are only referred by their hash (perhaps a fix is to directly read some meta/xml file containing non-hashed ped names or even recurse texture items).

This may be of help:
```c++
// int GET_VEHICLE_CLASS_FROM_NAME(Hash modelHash) // 0xDEDF1C8BD47C2200 0xEA469980 b323
// For a full enum, see here : pastebin.com/i2GGAjY0
// 
// char buffer[128];
// std::sprintf(buffer, "VEH_CLASS_%i", VEHICLE::GET_VEHICLE_CLASS_FROM_NAME (hash));
// 
// const char* className = HUD::_GET_LABEL_TEXT(buffer);
// 
// Full list of vehicles by DurtyFree: https://github.com/DurtyFree/gta-v-data-dumps/blob/master/vehicles.json
int __stdcall do_get_vehicle_class_from_name_impl(Hash modelHash)
{
  __int64 modelInfo; // r8
  int result; // eax
  int v3; // [rsp+40h] [rbp+18h] BYREF

  v3 = 0xFFFF;
  modelInfo = CStreaming::GetModelInfo(modelHash, &v3);
  result = 0;
  if ( v3 != 0xFFFF )
  {
    if ( modelInfo )
      return *(modelInfo + 0x548) & 0x1F;
  }
  return result;
}
```
If caching is required, the already existing `datBitBuffer` implementation should do the trick,

**Reply by Yimura** [reply]
Certain members in CVehicleModelInfo (or CBaseModelInfo) are only populated after calling REQUEST_MODEL which will cause some load I'm expecting.

Regarding the ped hash strings those will have to be collected from meta/xml files or even just scraped from scripts (Unsure if scripts have ped spawn names, I know they have vehicle spawn names in a huge switch case).

For caching we could just create a struct that holds the fields we wish to store and just dump those bytes directly to the file, no need to use datBitBuffer.

**Reply by FIying-Scotsman** [reply]
Just to add to @maybegreat48's post - Zorg93 released his "vehicle auto-updater" code a while ago which essentially goes to the entity pool, pulls out the vehicles and then sorts them into buckets depending on vehicle type. Menyoo has an implementation of it [here](https://github.com/MAFINS/MenyooSP/blob/b216d9448450a3832f8ef892c1827603ad55ea5a/Solution/source/Scripting/ModelNames.cpp#L156).

As it goes through the entity pool - it supports custom vehicles which you can get the name of like any other vehicle.

From experience you'll have to tweak the timing if you were to implement it however. For some reason it works flawlessly in the Scripthook but when used in BBV2 (and forks of) can sometimes not populate the giant array fast enough so needs to be manually re-run or works but unloading and reloading causes some of the vehicle categories to fail to work while the others do.

**Comment by xiaoxiao921** [comment]
if we have to read from xml / meta files or scrapping from scripts we may aswell:

- read everything from rage fiDevice, and construct a cache from that fully locally, i'm currently reading all of this from the fivem repo, if there is a way to query a path list of all the fiDevice path we need, we can construct the cache at any point instead of having to hook very early in the game loading and instead be able to update the cache at any point of dll injection
